// docusaurus-plugin-internaldocs-fb/lib/auth.js
"use strict";
/**
 * (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
 *
 * @format
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNeedToAuthenticateMessage = exports.storeToken = exports.getAuthUrl = exports.loadToken = void 0;
const fs_1 = require("fs");
const os_1 = require("os");
const path_1 = __importStar(require("path"));
const { mkdir, readFile, writeFile } = fs_1.promises;
const TOKEN_PATH = path_1.default.join((0, os_1.homedir)(), '.config/static-docs/auth-token');
const USER_R_W = 0o600;
async function loadToken() {
    if (process.env.CAT_TOKEN && process.env.CAT_APP) {
        return {
            type: 'cat',
            payload: process.env.CAT_TOKEN,
            app: process.env.CAT_APP,
        };
    }
    try {
        const token = await readFile(TOKEN_PATH);
        return { type: 'oauth', payload: token.toString() };
    }
    catch (e) {
        // No token found. Return null as the build may be able to continue.
        return null;
    }
}
exports.loadToken = loadToken;
function getAuthUrl() {
    return 'https://www.internalfb.com/intern/oauth/static-docs';
}
exports.getAuthUrl = getAuthUrl;
async function storeToken(token) {
    await mkdir((0, path_1.dirname)(TOKEN_PATH), { recursive: true });
    await writeFile(TOKEN_PATH, token, { mode: USER_R_W });
}
exports.storeToken = storeToken;
function getNeedToAuthenticateMessage() {
    return `

==================================================================================================

            ░█████╗░████████╗████████╗███████╗███╗░░██╗████████╗██╗░█████╗░███╗░░██╗
            ██╔══██╗╚══██╔══╝╚══██╔══╝██╔════╝████╗░██║╚══██╔══╝██║██╔══██╗████╗░██║
            ███████║░░░██║░░░░░░██║░░░█████╗░░██╔██╗██║░░░██║░░░██║██║░░██║██╔██╗██║
            ██╔══██║░░░██║░░░░░░██║░░░██╔══╝░░██║╚████║░░░██║░░░██║██║░░██║██║╚████║
            ██║░░██║░░░██║░░░░░░██║░░░███████╗██║░╚███║░░░██║░░░██║╚█████╔╝██║░╚███║
            ╚═╝░░╚═╝░░░╚═╝░░░░░░╚═╝░░░╚══════╝╚═╝░░╚══╝░░░╚═╝░░░╚═╝░╚════╝░╚═╝░░╚══╝


As you've added a new code snippet from another repo, we need to authenticate so we can fetch it.

Please visit this page to authenticate first before trying the build again:

${getAuthUrl()}

==================================================================================================

`;
}
exports.getNeedToAuthenticateMessage = getNeedToAuthenticateMessage;
