// internaldocs-fb-helpers/dist/uidocs.js
"use strict";
/**
 * (c) Facebook, Inc. and its affiliates. Confidential and proprietary.
 *
 * @format
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getApi = exports.docsets = void 0;
const bridge_1 = require("./bridge");
exports.docsets = {
    BLOKS_CORE: '887372105406659'
};
function getApi(args) {
    return __awaiter(this, void 0, void 0, function* () {
        const { name, framework, docset } = args;
        return yield bridge_1.call({
            module: 'uidocs',
            api: 'getApi',
            args: { name, framework, docset },
        });
    });
}
exports.getApi = getApi;
